// @up-im/apcheki_sdk/src/interfaces/iCatalog.ts
import { iBrand } from "./iBrand"
import { iQuestion } from "./iQuestion"
import { iSubstance } from "./iSubstance"
import { iSymptoms } from "./iSymptoms"

export interface iCatalog {
    //товар
    pub_date: string
    pub_date_original: string
    node: number
    title: string
    disclaimer: { title: string } | null//подзаголовок
    code: string
    image: string
    video?: string
    brand?: iBrand
    symptoms?: iSymptoms[]
    substance?: iSubstance[]
    //акция
    id: number
    shop_all: 1 | 0//акция во всех аптеках
    cashback_percent: boolean
    cashback_level_1: number
    cashback_level_2?: number
    cashback_level_3?: number
    presentation_title?: string//Название презентации
    presentation_text?: string//Описание презентации
    presentation?: string[]
    questions_title?: string//Название теста
    questions_text?: string//Описание теста
    questions?: iQuestion[]
    date_start: string
    date_end: string
    text: string
    text_full: string
    //промоблок
    promo_title: string
    promo_subtitle: string
    promo_image: string
    promo_disclaimer: string
    promo_text: string
}
