// @up-im/apcheki_sdk/src/interfaces/iPayment.ts
import { PaymentStatus, PaymentStatusKeys } from "./paymentStatus";
import { PaymentType, PaymentTypeKeys } from "./paymentType";
import { PayoutType, PayoutTypeKeys } from "./payoutType";


export interface iPayment {
    id: number
    reg_date: string//дата создания
    status: typeof PaymentStatus[PaymentStatusKeys]
    pay_type: typeof PaymentType[PaymentTypeKeys]//тип операции, начисление или выплата
    title: string//краткое описание для вывода пользователю

    //начисление
    receipt: number | null//чек
    receipt_product: number | null//товар чека
    catalog: number | null//id товара каталога
    sku: number | null//id акции
    shop: number | null//id аптеки
    cashback_level: 0 | 1 | 2 | 3//уровень кэшбека по которому произошло начисление
    max_cashback_level: null | 0 | 1 | 2 | 3 //макс. возможный уровень кэшбека по акции, поле есть только у начислений в чеках 

    //списание
    payout_type: typeof PayoutType[PayoutTypeKeys] | null //способ выплаты
    last_error: string//текст ошибки выплаты

    //размер
    price: number //если больше 0 начисление, меньше 0 списание
}
