// @up-im/apcheki_sdk/dist/api/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createClient = createClient;
const default_1 = require("../storage/default");
const user_1 = require("../storage/user");
const bookmarks_1 = require("./bookmarks");
const chat_1 = require("./chat");
const content_1 = require("./content");
const geoIp_1 = require("./geoIp");
const payment_1 = require("./payment");
const push_1 = require("./push");
const receipt_1 = require("./receipt");
const stat_1 = require("./stat");
const user_2 = require("./user");
function createClient({ host, wsHost, storage }) {
    const userStorage = new user_1.UserStorage(storage || default_1.storageDefault);
    return {
        user: new user_2.ApiUser({ host, userStorage }),
        content: new content_1.ApiContent({ host, userStorage }),
        geoIp: new geoIp_1.ApiGeoIp({ host, userStorage }),
        receipt: new receipt_1.ApiReceipt({ host, userStorage }),
        payment: new payment_1.ApiPayment({ host, userStorage }),
        stat: new stat_1.ApiStat({ host, userStorage }),
        chat: new chat_1.ApiChat({ host, wsHost, userStorage }),
        bookmarks: new bookmarks_1.ApiBookmarks({ host, userStorage }),
        push: new push_1.ApiPush({ host, userStorage }),
    };
}
