// @up-im/apcheki_sdk/src/api/index.ts
import { iStorage } from "../interfaces/iStorage";
import { storageDefault } from "../storage/default";
import { UserStorage } from "../storage/user";
import { ApiBookmarks } from "./bookmarks";
import { ApiChat } from "./chat";
import { ApiContent } from "./content";
import { ApiGeoIp } from "./geoIp";
import { ApiPayment } from "./payment";
import { ApiPush } from "./push";
import { ApiReceipt } from "./receipt";
import { ApiStat } from "./stat";
import { ApiUser } from "./user";

export function createClient({ host, wsHost, storage }: { host: string, wsHost: string, storage?: iStorage }) {
    const userStorage = new UserStorage(storage || storageDefault)
    return {
        user: new ApiUser({ host, userStorage }),
        content: new ApiContent({ host, userStorage }),
        geoIp: new ApiGeoIp({ host, userStorage }),
        receipt: new ApiReceipt({ host, userStorage }),
        payment: new ApiPayment({ host, userStorage }),
        stat: new ApiStat({ host, userStorage }),
        chat: new ApiChat({ host, wsHost, userStorage }),
        bookmarks: new ApiBookmarks({ host, userStorage }),
        push: new ApiPush({ host, userStorage }),
    }
}
